lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,159 @@ namespace Js
         return CreateThenPromise(promise, fulfillmentHandler, rejectionHandler, scriptContext);
     }
 
+    // Promise.prototype.finally as described in the draft of ES 2018 Section 25.4.5.3
+    Var JavascriptPromise::EntryFinally(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        ScriptContext* scriptContext = function->GetScriptContext();
+
+        AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("Promise.prototype.finally"));
+        //1. Let promise be the this value
+        //2. If Type(promise) is not Object, throw a TypeError exception
+        if (args.Info.Count < 1 || !JavascriptOperators::IsObject(args[0]))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NeedObject, _u("Promise.prototype.finally"));
+        }
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
```

Comment:
>UnsafeFromVar(args[0]); [](start = 54, length = 23)

Is it okay to call UnsafeFromVar(args[0]) here? -- do we know for sure this has to be a RecyclableObject at this point? It is possible to call with this as, say, a number?

Reply:
Or, because the check above covers that it is an object -- is it necessarily a RecyclableObject or could it be something else (I'm not sure the answer to this question)

---
In reply to: [167733902](https://github.com/Microsoft/ChakraCore/pull/4668#discussion_r167733902) [](ancestors = 167733902)

Reply:
We did a check above to make sure we have an args[0] and that it is an object. Should be guaranteed here.

Reply:
There's a type error thrown a couple of lines above if the arg is not an Object.

I'll also add a new test case that confirms that a type error is thrown if this is called with any non-promise object.